Content.Server/Shuttles/Systems/ShuttleSystem.Impact.cs

```diff
@@ -1,60 +1,398 @@
+            var ourTile = new Vector2i((int)Math.Floor(ourPoint.X / ourGrid.TileSize), (int)Math.Floor(ourPoint.Y / ourGrid.TileSize));
+            var otherTile = new Vector2i((int)Math.Floor(otherPoint.X / otherGrid.TileSize), (int)Math.Floor(otherPoint.Y / otherGrid.TileSize));
 
-        var ourVelocity = _physics.GetLinearVelocity(uid, ourPoint, ourBody, ourXform);
-        var otherVelocity = _physics.GetLinearVelocity(args.OtherEntity, otherPoint, otherBody, otherXform);
-        var jungleDiff = (ourVelocity - otherVelocity).Length();
+            var ourMass = GetRegionMass(uid, ourGrid, ourTile, ImpactRadius, out var ourTiles);
+            var otherMass = GetRegionMass(args.OtherEntity, otherGrid, otherTile, ImpactRadius, out var otherTiles);
 
-        if (jungleDiff < MinimumImpactVelocity)
+            if (ourTiles == 0 || otherTiles == 0) // i have no idea why this happens
+                continue;
+
+            Log.Info($"Shuttle impact of {ToPrettyString(uid)} with {ToPrettyString(args.OtherEntity)}; our mass: {ourMass}, other: {otherMass}, velocity {jungleDiff}, impact point {worldPoint}");
+
+            var energyMult = MathF.Pow(jungleDiff, 2) / 2;
+            // multiplier to make the area with more mass take less damage so a reinforced wall rammer doesn't die to lattice
+            var biasMult = MathF.Pow(ourMass / otherMass, MassBias);
+            // multiplier to make large grids not just bonk against each other
+            var inertiaMult = MathF.Pow(effectiveInertiaMult / BaseShuttleMass, InertiaScaling);
+            var ourEnergy = ourMass * energyMult * inertiaMult * MathF.Min(1f, biasMult);
+            var otherEnergy = otherMass * energyMult * inertiaMult / MathF.Max(1f, biasMult);
+
+            var ourRadius = Math.Min(ImpactRadius, MathF.Sqrt(otherEnergy / TileBreakEnergyMultiplier / PlatingMass));
+            var otherRadius = Math.Min(ImpactRadius, MathF.Sqrt(ourEnergy / TileBreakEnergyMultiplier / PlatingMass));
+
+            var totalInertia = ourVelocity * ourMass + otherVelocity * otherMass;
+            var unelasticVel = totalInertia / (ourMass + otherMass);
+            var ourPostImpactVelocity = Vector2.Lerp(ourVelocity, unelasticVel, MathF.Min(1f, ImpactSlowdown * ourTiles * args.OurFixture.Density / ourBody.FixturesMass));
+            var otherPostImpactVelocity = Vector2.Lerp(otherVelocity, unelasticVel, MathF.Min(1f, ImpactSlowdown * otherTiles * args.OtherFixture.Density / otherBody.FixturesMass));
+            var ourDeltaV = -ourVelocity + ourPostImpactVelocity;
+            var otherDeltaV = -otherVelocity + otherPostImpactVelocity;
+            _physics.ApplyLinearImpulse(uid, ourDeltaV * ourBody.FixturesMass, body: ourBody);
+            _physics.ApplyLinearImpulse(args.OtherEntity, otherDeltaV * otherBody.FixturesMass, body: otherBody);
+
+            var dir = (ourVelocity.Length() > otherVelocity.Length() ? ourVelocity : -otherVelocity).Normalized();
```

Comment:
```suggestion
            var dir = (ourVelocity.LengthSquared() > otherVelocity.LengthSquared() ? ourVelocity : -otherVelocity).Normalized();
```
very minor optimization